mailpoet/lib/Automation/Engine/Builder/DuplicateWorkflowController.php

```diff
@@ -0,0 +1,91 @@
+<?php declare(strict_types = 1);
+
+namespace MailPoet\Automation\Engine\Builder;
+
+use MailPoet\Automation\Engine\Data\NextStep;
+use MailPoet\Automation\Engine\Data\Step;
+use MailPoet\Automation\Engine\Data\Workflow;
+use MailPoet\Automation\Engine\Exceptions;
+use MailPoet\Automation\Engine\Exceptions\InvalidStateException;
+use MailPoet\Automation\Engine\Storage\WorkflowStorage;
+use MailPoet\Automation\Engine\WordPress;
+use MailPoet\Util\Security;
+
+class DuplicateWorkflowController {
+  /** @var WordPress */
+  private $wordPress;
+
+  /** @var WorkflowStorage */
+  private $workflowStorage;
+
+  public function __construct(
+    WordPress $wordPress,
+    WorkflowStorage $workflowStorage
+  ) {
+    $this->wordPress = $wordPress;
+    $this->workflowStorage = $workflowStorage;
+  }
+
+  public function duplicateWorkflow(int $id): Workflow {
+    $workflow = $this->workflowStorage->getWorkflow($id);
+    if (!$workflow) {
+      throw Exceptions::workflowNotFound($id);
+    }
+
+    $duplicate = new Workflow(
+      $this->getName($workflow->getName()),
+      $this->getSteps($workflow->getSteps()),
+      $this->wordPress->wpGetCurrentUser()
+    );
+    $duplicate->setStatus(Workflow::STATUS_DRAFT);
+
+    $workflowId = $this->workflowStorage->createWorkflow($duplicate);
+    $savedWorkflow = $this->workflowStorage->getWorkflow($workflowId);
+    if (!$savedWorkflow) {
+      throw new InvalidStateException('Workflow not found.');
+    }
+    return $savedWorkflow;
+  }
+
+  private function getName(string $name): string {
+    // translators: %s is the original workflow name.
+    $newName = sprintf(__('Copy of %s', 'mailpoet'), $name);
+    $maxLength = $this->workflowStorage->getNameColumnLength();
+    if (strlen($newName) > $maxLength) {
+      $append = '…';
+      return substr($newName, 0, $maxLength - strlen($append)) . $append;
+    }
+    return $newName;
+  }
+
+  /**
+   * @param Step[] $steps
+   * @return Step[]
+  */
+  private function getSteps(array $steps): array {
+    $newIds = [];
+    foreach ($steps as $step) {
+      $id = $step->getId();
+      $newIds[$id] = $id === 'root' ? 'root' : $this->getId();
```

Comment:
Out of curiosity: Do we generate new IDs only because, they are IDs and should be unique or is there a material reason for it I do miss? Or, to rephrase: Isn't it enough to be sure the IDs are unique in the same workflow.

If it is about: Maybe some day, we want to extract steps into its own table, I think we would need to check the uniqueness at that point anyway (if we would want to use this `getID()`-string in a `unique` column, given `root`).

I just wonder, if we could get rid of all that logic and just use the original `Step[]`s, especially as we return `$savedWorkflow` anyway, so the consumer of this method will get a copy and not the original `Step` objects.

Reply:
@websupporter I added a [note to myself](https://github.com/mailpoet/mailpoet/pull/4376#discussion_r994535062) when going through the code:

> I'm wondering if duplicating a workflow should also change step IDs. It's not necessary (we never said they are universally unique) but it might be clearer and better for debugging, maybe 🤔

You're right that this is not necessary, but my thinking was more like "just to be sure" not to mix up those two workflows somewhere together and make it clear for debugging, etc. Essentially, I wanted to prevent making duplicate workflow IDs explicitly (although we don't consider them universally unique).

Reply:
Ok :+1: